### PR TITLE
8341595: Clean up iteration of CEN headers in ZipFile.Source.initCEN

### PR DESCRIPTION
**Pull request by eirbjo**
Please review this PR which suggests we clean up iteration and validation logic in `ZipFile.Source::initCEN` and some related methods to use a simpler and more consistent style:

* The main loop in `ZipFile.Source::initCEN` currently updates two iteration variables (`pos` and `entryPos`), where `entryPos` is always just `pos + CENHDR`. One variable should be sufficient. `entryPos` can be moved to an effectively final local variable scoped inside the loop.
* The local variable `int limit = cen.length` no longer carries its weight and is inlined into its only use site
* Since `entryPos` is no longer in scope for the loop predicate, this is updated to `pos <= cen.length - CENHDR`, instead of the current `entryPos <= limit`
* The byte array passed to `countCENHeaders` contains only CEN data now, so the `size` parameter can be removed.
*  `countCENHeaders` is updated to loop on the same predicate as `initCEN`. Additionally, this method is updated to  throw early if a CEN entry exceeds the CEN size (for consistency with similar logic in `checkAndAddEntry`)
* The `headerSize` validation in `checkAndAddEntry` is updated to avoid widening conversion to `long` of a variable which can provably never exceed `Integer.MAX_VALUE` and to be consistent with  `countCENHeaders`.

Testing: 

A new test `CenSizeMaximum` is added:
* This produces a ZIP file with a CEN size of exactly `MAX_CEN_SIZE` and verifies that the iteration logic handles ZIP files where the CEN size is on or near the limit.
* This test also manipulates the END headers  'total number of entries' field, in order to exercise `countCENHeader`. 
* The test is a bit of a resource hog: It produces a  ~1.5GB ZIP file on disk, requires > 2GB heap and takes > 15 seconds to run on my laptop. Let me know if this should be made a manual test.

GHA tests are currently pending. I have run the following tests locally:

```
% make test TEST="test/jdk/java/util/zip" 
% make test TEST="test/jdk/java/util/jar" 
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341595](https://bugs.openjdk.org/browse/JDK-8341595): Clean up iteration of CEN headers in ZipFile.Source.initCEN (**Enhancement** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) Review applies to [142bc7ac](https://git.openjdk.org/jdk/pull/21378/files/142bc7ac427fce26f81af4f08b34d1f339cac460)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21378/head:pull/21378` \
`$ git checkout pull/21378`

Update a local copy of the PR: \
`$ git checkout pull/21378` \
`$ git pull https://git.openjdk.org/jdk.git pull/21378/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21378`

View PR using the GUI difftool: \
`$ git pr show -t 21378`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21378.diff">https://git.openjdk.org/jdk/pull/21378.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21378#issuecomment-2396626771)

### CODE REVIEW

File: src/java.base/share/classes/java/util/zip/ZipFile.java [hunk 1]
```diff
@@ -1768,18 +1768,17 @@ private void initCEN(int knownTotal) throws IOException {
             // Iterate through the entries in the central directory
             int idx = 0; // Index into the entries array
             int pos = 0;
-            int entryPos = CENHDR;
-            int limit = cen.length;
             manifestNum = 0;
-            while (entryPos <= limit) {
+            while (pos <= cen.length - CENHDR) {
```
Comment (cl4es) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1772]: The interpreter would probably like something like: 
```
int limit = cen.length - CENHDR;
while (pos <= limit) {
```

File: test/jdk/java/util/zip/ZipFile/CenSizeMaximum.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
```
Comment (cl4es) [inline test/jdk/java/util/zip/ZipFile/CenSizeMaximum.java:24]: Yes, this test probably should be either manual or moved to a higher tier.

File: src/java.base/share/classes/java/util/zip/ZipFile.java (hunk 1 above)
Comment (eirbjo) [reply src/java.base/share/classes/java/util/zip/ZipFile.java]: Fixed, look like a good improvement.

File: test/jdk/java/util/zip/ZipFile/CenSizeMaximum.java (hunk 2 above)
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/CenSizeMaximum.java:24]: Added `/manual` for now since I wouldn't know how to move tests between tiers :-)

**Review by LanceAndersen — APPROVED** [review]
Thank you for the clean up Eirik... 

looks good

**Pull request closed without merging**